python/ql/src/experimental/Security-new-dataflow/CWE-022/PathInjection.ql

```diff
@@ -0,0 +1,112 @@
+ *
+ *  2. The path is normalized at least once, but never checked afterwards.
+ *     This is detected by finding the first normalization and then ensure that
+ *     no checks happen after. Since we start from the first normalization,
+ *     we know that the absence of checks means that no normalization has a
+ *     chek after it. (No checks after a second normalization would be ok if
+ *     there was a check between the first and the second.)
+ *
+ * Note that one could make the dual split on whether the path is ever checked. This does
+ * not work as nicely, however, since checking is modelled as a `BarrierGuard` rather than
+ * as a `Sanitizer`. That means that only some paths out of a check will be removed, and so
+ * identifying the last check is not possible simply by finding a path from it to a sink.
+ *
+ * @name Uncontrolled data used in path expression
+ * @description Accessing paths influenced by users can allow an attacker to access unexpected resources.
+ * @kind path-problem
+ * @problem.severity error
+ * @sub-severity high
+ * @precision high
+ * @id py/path-injection
+ * @tags correctness
+ *       security
+ *       external/owasp/owasp-a1
+ *       external/cwe/cwe-022
+ *       external/cwe/cwe-023
+ *       external/cwe/cwe-036
+ *       external/cwe/cwe-073
+ *       external/cwe/cwe-099
+ */
+
+import python
+import experimental.dataflow.DataFlow
+import experimental.dataflow.DataFlow2
+import experimental.dataflow.TaintTracking
+import experimental.dataflow.TaintTracking2
+import experimental.semmle.python.Concepts
+import experimental.dataflow.RemoteFlowSources
+import ChainedConfigs12
+
+// ---------------------------------------------------------------------------
+// Case 1. The path is never normalized.
+// ---------------------------------------------------------------------------
+/** Configuration to find paths from sources to sinks that contain no normalization. */
+class PathNotNormalizedConfiguration extends TaintTracking::Configuration {
+  PathNotNormalizedConfiguration() { this = "PathNotNormalizedConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sink = any(FileSystemAccess e).getAPathArgument()
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) { node instanceof PathNormalization }
+}
+
+predicate pathNotNormalized(CustomPathNode source, CustomPathNode sink) {
+  any(PathNotNormalizedConfiguration config).hasFlowPath(source.asNode1(), sink.asNode1())
+}
+
+// ---------------------------------------------------------------------------
+// Case 2. The path is normalized at least once, but never checked afterwards.
+// ---------------------------------------------------------------------------
+/** Configuration to find paths from sources to normalizations that contain no prior normalizations. */
+class FirstNormalizationConfiguration extends TaintTracking::Configuration {
```

Comment:
Probably, they represent disjoint sets of paths. We could use a `TaintTracking3::Configuration` to be sure?

Reply:
Asked on slack:
> Is it safe in general to have two configurations extending TaintTracking::Configuration if they represent disjoint sets of paths?

got answer (from `@rdmarsh2`):
> No
There's performance issues with it
Also it will create negative recursion if overrides in one configuration depend on the hasFlow from the other
No correctness issues that I'm aware of, though

So I have added `TaintTracking3/4` (and the then needed `DataFlow3/4`). 4 will be needed if we make the rewrite suggested by `@aschackmull`. There is now a test showing the benefit of that rewrite.

Reply:
Got additional answer from `@p0`:
> Crucially, I'd emphasise "there can be a performance cost". It's by no means automatic that two subtypes of `DataFlow::Configuration` would be slower than one subtyping `DataFlow::Configuration` and the other `DataFlow2::Configuration`.

> The rule of thumb I'd apply is to work with a single configuration supertype, unless the compiler prevents you from doing so (because your configurations are mutually recursive in nonmonotonic ways), or you actually observe some blow-up that's caused by unrelated configurations (that'd be a relatively specialised situation, I think).

Reply:
And a comment from `@aschackmull`:
> FWIW, the java libs do indeed use multiple configurations deriving from the same library copy.
Also, if you're using data flow inside some library code that's imported into multiple queries then try to avoid using the first data-flow copy in those cases, since one way to be certain to get a performance cost is to recalculate some library data flow each time some query creates a new sibling configuration (and query writers tend to use the first copy by default).

Reply:
Using the same configuration allow us to be part of the same path graph, which allows the nice formulation of the query currently seen.